### PR TITLE
Sidecar utilities

### PR DESCRIPTION
**Pull request by alhom**
A bunch of utilities to calculate differential ops for sidecar files.

### CODE REVIEW

File: src/Vlasiator.jl [hunk 1]
```diff
@@ -14,6 +14,8 @@ using Dates
 
 include("utility/rotation.jl")
 include("utility/log.jl")
+#include("utility/vector.jl")
+include("utility/differential.jl")
```
Comment (henry2004y) [inline src/Vlasiator.jl:18]: The old vector.jl has been renamed to array.jl in master. But maybe indeed differential.jl is a better name!

File: src/Vlasiator.jl [hunk 2]
```diff
@@ -35,14 +37,21 @@ export
...
    check_plasma_characteristics,
    # fluxfunction
    compute_flux_function, find_reconnection_points
 
+
 precompile(load, (String,))
 precompile(readvariable, (MetaVLSV, String, Bool))
 
+# oo, a typedef!
+realtype=Float64
```
Comment (henry2004y) [inline src/Vlasiator.jl:53]: Where is this `realtype` used?

File: src/vlsv/vlsvreader.jl [hunk 3]
```diff
@@ -317,7 +321,7 @@ Return variable value of `var` from the VLSV file associated with `meta`. By def
 function readvariable(meta::MetaVLSV, var, sorted::Bool=true)
    (;fid, footer, cellindex) = meta
    if (local symvar = Symbol(var)) in keys(variables_predefined)
-      data = variables_predefined[symvar](meta)
+      data = variables_predefined[symvar](meta, sorted)
```
Comment (henry2004y) [inline src/vlsv/vlsvreader.jl:324]: I guess this is currently not supported for the derived variables?

File: src/vlsv/vlsvreader.jl [hunk 4]
```diff
@@ -242,10 +243,13 @@ function load(file::AbstractString)
       !isempty(vcells)
    end
 
+   # Get fsgrid bbox
+   fsbox = NTuple{6}(readmesh(fid, footer, "fsgrid", "MESH_BBOX")::Vector{Int})
+   fsbox = NTuple{3}(fsbox[1:3])
```
Comment (henry2004y) [inline src/vlsv/vlsvreader.jl:248]: Do you need this `fsgridcells` anywhere? It is actually equivalent to `meta.ncells .* 2 .^meta.maxamr`

File: src/vlsv/vlsvutility.jl [hunk 5]
```diff
@@ -1170,4 +1170,25 @@ function issame(f1, f2, tol::AbstractFloat=1e-4; verbose=false)
...
+   downsamplefg(meta, array, ndims) -> vg array
+
+Downsamples a fg array (compatible with meta) of a variable with ndims components to SpatialGrid of the file meta.
+Returns the downsampled values in an array sized [length(meta.cellid), ndims].
+"""
+function downsample_fg(meta, arr, ndims)
+   if ndims > 1
+      vgout = zeros(typeof(arr[1]), (ndims,length(meta.cellid)))
+      for (i,cid) in enumerate(meta.cellid)
+         vgout[:,i] = downsample_fsgrid_subarray(meta, cid, arr)
```
Comment (henry2004y) [inline src/vlsv/vlsvutility.jl:1185]: Where is `downsample_fsgrid_subarray` defined?

File: src/vlsv/vlsvvariables.jl [hunk 6]
```diff
@@ -161,7 +162,7 @@ const variables_predefined = Dict(
       VS = @. √( (P*5.0f0/3.0f0) / ρm )
    end,
    :VA => function (meta, ids=UInt64[]) # Alfvén speed
-      ρm = readvariable(meta, "Rhom", ids)
+      ρm = readvariable(meta, :Rhom, ids)
```
Comment (henry2004y) [inline src/vlsv/vlsvvariables.jl:165]: Currently I still use strings, even though they are internally converted into symbols.

File: src/vlsv/vlsvvariables.jl [hunk 7]
```diff
@@ -73,10 +73,11 @@ const units_predefined = Dict(
...
    :Pperp => ("Pa", L"$P_\perp$", "Pa"),
-   :Gyroperiod => ("s", L"$T_{gyro}$", "s"),
-   :Plasmaperiod => ("s", L"$T_{plasma}$", "s"),
-   :Gyrofrequency => ("/s", L"$\omega_{g}$", L"s^{-1}"),
-   :Omegap => ("/s", L"$\omega_{p}$", L"s^{-1}"),
+   :Beta => ("",L"$\beta$", "")
+#   :Gyroperiod => ("s", L"$T_{gyro}$", "s"),
+#   :Plasmaperiod => ("s", L"$T_{plasma}$", "s"),
+#   :Gyrofrequency => ("/s", L"$\omega_{g}$", L"s^{-1}"),
+#   :Omegap => ("/s", L"$\omega_{p}$", L"s^{-1}"),
```
Comment (henry2004y) [inline src/vlsv/vlsvvariables.jl:80]: Why are they commented out?

File: src/utility/differential.jl [hunk 8]
```diff
@@ -0,0 +1,287 @@
+"""
```
Comment (henry2004y) [inline src/utility/differential.jl:1]: I haven't looked into this carefully, but I fixed some previous bugs in the original vector.jl based on the changes here in array.jl and adjusted the APIs a bit. Maybe we need to merge them together.

**Review by henry2004y — COMMENTED** [review]
Some tests failed, but I haven't looked into those. It would be great if you can provide some small tests with the fsgrid functionalities. Currently I have 3 small VLSV files used for testing purposes. If you think that's not enough to cover your cases, we can add another one or replace one of the existing files.

File: src/vlsv/vlsvutility.jl (hunk 5 above)
Comment (alhom) [reply src/vlsv/vlsvutility.jl:1185]: Looks like I have not moved that one from the initial workbench to the codebase. Derp.

File: src/vlsv/vlsvvariables.jl (hunk 7 above)
Comment (alhom) [reply src/vlsv/vlsvvariables.jl:80]: I think that was related to the sorted keyword not working, which I was trying to use at the time..

File: src/vlsv/vlsvreader.jl (hunk 4 above)
Comment (alhom) [reply src/vlsv/vlsvreader.jl:248]: Yes, in the portion that I hadn't moved yet to vlsvutility.jl... Also, TBH, `fsgridcells` is more readable than `meta.ncells .* 2 .^meta.maxamr` :)

File: src/Vlasiator.jl (hunk 2 above)
Comment (alhom) [reply src/Vlasiator.jl:53]: Nowhere. Removed.

File: src/vlsv/vlsvvariables.jl (hunk 7 above)
Comment (alhom) [reply src/vlsv/vlsvvariables.jl:80]: Ha, no! Uncommenting them leads to a compilation error (Variable not defined)... which was the thing I was scratching my head at. Now I noticed the root cause for that - I had added them _without_ adding a comma above to the list :man_facepalming:

File: src/vlsv/vlsvreader.jl (hunk 3 above)
Comment (alhom) [reply src/vlsv/vlsvreader.jl:324]: Right, no. Adding that functionality went bust with the comma derp I just found, and that is a leftover from that.

**Comment by alhom** [comment]
Fixed a bunch of stuff and brought in the missing parts of utilities. `differential.jl` is still as its own self, I'll let you decide how to proceed with that, and no test cases yet.

File: src/vlsv/vlsvvariables.jl (hunk 7 above)
Comment (henry2004y) [reply src/vlsv/vlsvvariables.jl]: That's fine. I can handle these afterwards.

**Pull request closed without merging**